### DISCUSSION TITLE
Potential bugs/feature to enhance & queries?

### DISCUSSION THREAD

**Discussion by skwskwskwskw** [deepdoctection/deepdoctection#106] [Q&A] [answered]
Some queries when going through the pipelines: 

1. How to read local image in pipeline? 

When using pipe.analyze for image file (with extension *.png or *.jpg); it returns error; but it works completely fine for *.pdf file (though only one page is read); 

![image](https://user-images.githubusercontent.com/7208012/209800642-27c7f65d-32c1-42af-944f-a3ccc1976ae3.png)

2. On top of this, possible to loop through every page in *.pdf file? 

3. Was using this HTML(table.html) - returning some results I am anticipating in notebook based on the preview - but when switching it to pandas table using pd.read_html - the table consists of only single column. Is there a way to obtain pandas table structure like preview in HTML?

4. Can I be sure the pipelines can detect more than 1 table in a page?

**Comment by jayeshthukarul** [comment, marked as answer]
Pipeline does extraction for all pdf pages.
4. Pipeline detects more than 1 table from the same page. It's totally depend on the performance of layout analysis model. You might need to finetune the model for your use case.

**Reply by skwskwskwskw** [reply]
> Pipeline does extraction for all pdf pages. 

When I run on multiple pdf pages, I got one page extracted only - where's the mistake I make? Can you share an example?

**Reply by jayeshthukarul** [reply]
I need to see your pdf example or the code how you are processing the pdf. Then I guess I will have more clearity and will be able to help you.

**Reply by skwskwskwskw** [reply]
It was pretty simple - just using the following: 

analyzer = dd.get_dd_analyzer(language='eng')
df = analyzer. Analyze(path="XXX.pdf")

It only detects the first page; 
Also, the screenshot above contained custom pipelines. 

I realised that in one of the tutorial Dr @JaMe76 mentioned it can only extract image from first page; kindly correct me if I am wrong.

**Reply by jayeshthukarul** [reply]
Ok. The code from attached screenshot above has issues. Give me some time. I will post correct code. Meanwhile you can correct the error shown above by providing the full path of the image file instead of file name. If the file is available in the same directory as the code file provide path like './filename.jpg'

**Reply by JaMe76** [reply]
The analyzer processes images all contained in one dir or all pages of a single pdf-file. You can therefore pass only a path to a **directory** that contains images or a single file to a pdf with maybe multiple pages. You cannot pass only a single image or a directory of multiple pdfs.

Regarding your code above if you want to process all pages of one pdf you must call `next` multiple times or even simpler you must loop:

``` 
import deepdoctection as dd

analyzer = dd.get_dd_analyzer(language='eng')
df = analyzer.analyze(path="path/to/pdf")
df.reset_state()
for dp in df:
   dp.viz(interactive=True).  # or something similar
```

If you want to process multiple pdfs you can either merge them into one file or setup a stream of multiple pdfs like that:

```
import deepdoctection as dd

pdf_path_list = ["path/to/pdf1","path/to/pdf1"]
analyzer = dd.get_dd_anylzer()

df_list = []
for pdf_path in pdf_path_list:
    df_list.append(analyzer.analyze(path=pdf_path) setting up a dataflow for each pdf
 
df = dd.ConcatData(df_list).  # when one dataflow of one pdf is exhausted will start the next
df.reset_state()
for dp in df:
   ... # now will process all pages of each pdf. 
```

This answers 1.) and 2.)

3.) I do not know how pandas treats HTML internally but one problem might occur if your HTML has a cell spanning over two columns. As far as I know Pandas does not have the concept of spanning cells. If you want to pass a table with spanning cells you must decide to which column you want to assign the cell (one column/all columns).
